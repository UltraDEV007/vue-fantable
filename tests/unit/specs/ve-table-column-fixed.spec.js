import { mount } from "@vue/test-utils";
import veTable from "@/ve-table";

describe("veTable column fixed", () => {
    const TABLE_DATA = [
        {
            col1: "1",
            col2: "2",
            col3: "3",
            col4: "4",
            col5: "5",
            col6: "6",
            col7: "7",
            col8: "8",
            col9: "9",
            col10: "10",
        },
        {
            col1: "1",
            col2: "2",
            col3: "3",
            col4: "4",
            col5: "5",
            col6: "6",
            col7: "7",
            col8: "8",
            col9: "9",
            col10: "10",
        },
        {
            col1: "1",
            col2: "2",
            col3: "3",
            col4: "4",
            col5: "5",
            col6: "6",
            col7: "7",
            col8: "8",
            col9: "9",
            col10: "10",
        },
        {
            col1: "1",
            col2: "2",
            col3: "3",
            col4: "4",
            col5: "5",
            col6: "6",
            col7: "7",
            col8: "8",
            col9: "9",
            col10: "10",
        },
        {
            col1: "1",
            col2: "2",
            col3: "3",
            col4: "4",
            col5: "5",
            col6: "6",
            col7: "7",
            col8: "8",
            col9: "9",
            col10: "10",
        },
    ];

    const COLUMNS = [
        {
            field: "col1",
            key: "a",
            title: "Title1",
            fixed: "left",
        },
        {
            field: "col2",
            key: "b",
            title: "Title2",
            fixed: "left",
        },
        {
            field: "col3",
            key: "c",
            title: "Title3",
        },
        { field: "col4", key: "d", title: "Title4" },
        { field: "col5", key: "e", title: "Title5" },
        { field: "col6", key: "f", title: "Title6" },
        {
            field: "col7",
            key: "g",
            title: "Title7",
        },
        {
            field: "col8",
            key: "h",
            title: "Title8",
        },
        {
            field: "col9",
            key: "i",
            title: "Title9",
            fixed: "right",
        },
        {
            field: "col10",
            key: "j",
            title: "Title10",
            fixed: "right",
        },
    ];

    // 固定列改变
    const COLUMNS_FIXED_CHANGE = [
        {
            field: "col1",
            key: "a",
            title: "Title1",
            fixed: "left",
        },
        {
            field: "col2",
            key: "b",
            title: "Title2",
        },
        {
            field: "col3",
            key: "c",
            title: "Title3",
        },
        { field: "col4", key: "d", title: "Title4" },
        { field: "col5", key: "e", title: "Title5" },
        { field: "col6", key: "f", title: "Title6" },
        {
            field: "col7",
            key: "g",
            title: "Title7",
        },
        {
            field: "col8",
            key: "h",
            title: "Title8",
        },
        {
            field: "col9",
            key: "i",
            title: "Title9",
        },
        {
            field: "col10",
            key: "j",
            title: "Title10",
            fixed: "right",
        },
    ];

    it("renders normal", () => {
        const wrapper = mount({
            render() {
                return <veTable columns={COLUMNS} tableData={TABLE_DATA} />;
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    it("header fixed left", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: COLUMNS,
                tableData: TABLE_DATA,
            },
        });

        const thEls = wrapper.findAll(
            ".ve-table-header .ve-table-header-tr .ve-table-header-th",
        );

        expect(thEls.at(0).classes()).toContain("ve-table-fixed-left");
        expect(thEls.at(1).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-fixed-left",
                "ve-table-last-left-fixed-column",
            ]),
        );

        await wrapper.setProps({ columns: COLUMNS_FIXED_CHANGE });

        const thEls2 = wrapper.findAll(
            ".ve-table-header .ve-table-header-tr .ve-table-header-th",
        );

        expect(thEls2.at(0).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-fixed-left",
                "ve-table-last-left-fixed-column",
            ]),
        );
    });

    it("header fixed right", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: COLUMNS,
                tableData: TABLE_DATA,
            },
        });

        const thEls = wrapper.findAll(
            ".ve-table-header .ve-table-header-tr .ve-table-header-th",
        );

        expect(thEls.at(9 - 1).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-first-right-fixed-column",
                "ve-table-fixed-right",
            ]),
        );
        expect(thEls.at(9 - 0).classes()).toContain("ve-table-fixed-right");

        await wrapper.setProps({ columns: COLUMNS_FIXED_CHANGE });

        const thEls2 = wrapper.findAll(
            ".ve-table-header .ve-table-header-tr .ve-table-header-th",
        );

        expect(thEls2.at(9 - 0).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-first-right-fixed-column",
                "ve-table-fixed-right",
            ]),
        );
    });

    it("column fixed left", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: COLUMNS,
                tableData: TABLE_DATA,
            },
        });

        const tdEls = wrapper
            .findAll(".ve-table-body .ve-table-body-tr")
            .at(0)
            .findAll(".ve-table-body-td");

        expect(tdEls.at(0).classes()).toContain("ve-table-fixed-left");
        expect(tdEls.at(1).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-last-left-fixed-column",
                "ve-table-fixed-left",
            ]),
        );

        await wrapper.setProps({ columns: COLUMNS_FIXED_CHANGE });

        const tdEls2 = wrapper
            .findAll(".ve-table-body .ve-table-body-tr")
            .at(0)
            .findAll(".ve-table-body-td");

        expect(tdEls2.at(0).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-last-left-fixed-column",
                "ve-table-fixed-left",
            ]),
        );
    });

    it("column fixed right", async () => {
        const wrapper = mount(veTable, {
            propsData: {
                columns: COLUMNS,
                tableData: TABLE_DATA,
            },
        });

        const tdEls = wrapper
            .findAll(".ve-table-body .ve-table-body-tr")
            .at(0)
            .findAll(".ve-table-body-td");

        expect(tdEls.at(9 - 1).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-first-right-fixed-column",
                "ve-table-fixed-right",
            ]),
        );
        expect(tdEls.at(9 - 0).classes()).toContain("ve-table-fixed-right");

        await wrapper.setProps({ columns: COLUMNS_FIXED_CHANGE });

        const tdEls2 = wrapper
            .findAll(".ve-table-body .ve-table-body-tr")
            .at(0)
            .findAll(".ve-table-body-td");

        expect(tdEls.at(9 - 0).classes()).toEqual(
            expect.arrayContaining([
                "ve-table-first-right-fixed-column",
                "ve-table-fixed-right",
            ]),
        );
    });
});
