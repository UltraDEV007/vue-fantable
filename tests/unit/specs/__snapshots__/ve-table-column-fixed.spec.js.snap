// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FanTable column fixed renders normal 1`] = `
<div class="vue-table-root">
  <div class="fan-table fan-table-border-around" style="width: 100%;">
    <div class="fan-table-container" style="height: 0px;">
      <div class="fan-table-content-wrapper">
        <table class="fan-table-content fan-table-border-x">
          <colgroup>
            <col>
            <col>
            <col>
            <col>
            <col>
            <col>
            <col>
            <col>
            <col>
            <col>
          </colgroup>
          <thead class="fan-table-fixed-header fan-table-header">
            <tr class="fan-table-header-tr">
              <th rowspan="1" colspan="1" class="fan-table-header-th fan-table-fixed-left" style="text-align: center; left: 0px; top: 0px;">Title1</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th fan-table-fixed-left fan-table-last-left-fixed-column" style="text-align: center; top: 0px;">Title2</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: center; top: 0px;">Title3</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: center; top: 0px;">Title4</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: center; top: 0px;">Title5</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: center; top: 0px;">Title6</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: center; top: 0px;">Title7</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: center; top: 0px;">Title8</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th fan-table-fixed-right fan-table-first-right-fixed-column" style="text-align: center; top: 0px;">Title9</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th fan-table-fixed-right fan-table-last-column" style="text-align: center; right: 0px; top: 0px;">Title10</th>
            </tr>
          </thead>
          <tbody class="fan-table-body fan-table-row-hover fan-table-row-highlight">
            <tr style="height: 0px;">
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
            </tr>
            <tr class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="a" class="fan-table-body-td fan-table-fixed-left" style="text-align: center; left: 0px;">1</td>
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td fan-table-fixed-left fan-table-last-left-fixed-column" style="text-align: center;">2</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: center;">3</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: center;">4</td>
              <td rowspan="1" colspan="1" col-key="e" class="fan-table-body-td" style="text-align: center;">5</td>
              <td rowspan="1" colspan="1" col-key="f" class="fan-table-body-td" style="text-align: center;">6</td>
              <td rowspan="1" colspan="1" col-key="g" class="fan-table-body-td" style="text-align: center;">7</td>
              <td rowspan="1" colspan="1" col-key="h" class="fan-table-body-td" style="text-align: center;">8</td>
              <td rowspan="1" colspan="1" col-key="i" class="fan-table-body-td fan-table-fixed-right fan-table-first-right-fixed-column" style="text-align: center;">9</td>
              <td rowspan="1" colspan="1" col-key="j" class="fan-table-body-td fan-table-fixed-right" style="text-align: center; right: 0px;">10</td>
            </tr>
            <tr class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="a" class="fan-table-body-td fan-table-fixed-left" style="text-align: center; left: 0px;">1</td>
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td fan-table-fixed-left fan-table-last-left-fixed-column" style="text-align: center;">2</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: center;">3</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: center;">4</td>
              <td rowspan="1" colspan="1" col-key="e" class="fan-table-body-td" style="text-align: center;">5</td>
              <td rowspan="1" colspan="1" col-key="f" class="fan-table-body-td" style="text-align: center;">6</td>
              <td rowspan="1" colspan="1" col-key="g" class="fan-table-body-td" style="text-align: center;">7</td>
              <td rowspan="1" colspan="1" col-key="h" class="fan-table-body-td" style="text-align: center;">8</td>
              <td rowspan="1" colspan="1" col-key="i" class="fan-table-body-td fan-table-fixed-right fan-table-first-right-fixed-column" style="text-align: center;">9</td>
              <td rowspan="1" colspan="1" col-key="j" class="fan-table-body-td fan-table-fixed-right" style="text-align: center; right: 0px;">10</td>
            </tr>
            <tr class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="a" class="fan-table-body-td fan-table-fixed-left" style="text-align: center; left: 0px;">1</td>
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td fan-table-fixed-left fan-table-last-left-fixed-column" style="text-align: center;">2</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: center;">3</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: center;">4</td>
              <td rowspan="1" colspan="1" col-key="e" class="fan-table-body-td" style="text-align: center;">5</td>
              <td rowspan="1" colspan="1" col-key="f" class="fan-table-body-td" style="text-align: center;">6</td>
              <td rowspan="1" colspan="1" col-key="g" class="fan-table-body-td" style="text-align: center;">7</td>
              <td rowspan="1" colspan="1" col-key="h" class="fan-table-body-td" style="text-align: center;">8</td>
              <td rowspan="1" colspan="1" col-key="i" class="fan-table-body-td fan-table-fixed-right fan-table-first-right-fixed-column" style="text-align: center;">9</td>
              <td rowspan="1" colspan="1" col-key="j" class="fan-table-body-td fan-table-fixed-right" style="text-align: center; right: 0px;">10</td>
            </tr>
            <tr class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="a" class="fan-table-body-td fan-table-fixed-left" style="text-align: center; left: 0px;">1</td>
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td fan-table-fixed-left fan-table-last-left-fixed-column" style="text-align: center;">2</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: center;">3</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: center;">4</td>
              <td rowspan="1" colspan="1" col-key="e" class="fan-table-body-td" style="text-align: center;">5</td>
              <td rowspan="1" colspan="1" col-key="f" class="fan-table-body-td" style="text-align: center;">6</td>
              <td rowspan="1" colspan="1" col-key="g" class="fan-table-body-td" style="text-align: center;">7</td>
              <td rowspan="1" colspan="1" col-key="h" class="fan-table-body-td" style="text-align: center;">8</td>
              <td rowspan="1" colspan="1" col-key="i" class="fan-table-body-td fan-table-fixed-right fan-table-first-right-fixed-column" style="text-align: center;">9</td>
              <td rowspan="1" colspan="1" col-key="j" class="fan-table-body-td fan-table-fixed-right" style="text-align: center; right: 0px;">10</td>
            </tr>
            <tr class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="a" class="fan-table-body-td fan-table-fixed-left" style="text-align: center; left: 0px;">1</td>
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td fan-table-fixed-left fan-table-last-left-fixed-column" style="text-align: center;">2</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: center;">3</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: center;">4</td>
              <td rowspan="1" colspan="1" col-key="e" class="fan-table-body-td" style="text-align: center;">5</td>
              <td rowspan="1" colspan="1" col-key="f" class="fan-table-body-td" style="text-align: center;">6</td>
              <td rowspan="1" colspan="1" col-key="g" class="fan-table-body-td" style="text-align: center;">7</td>
              <td rowspan="1" colspan="1" col-key="h" class="fan-table-body-td" style="text-align: center;">8</td>
              <td rowspan="1" colspan="1" col-key="i" class="fan-table-body-td fan-table-fixed-right fan-table-first-right-fixed-column" style="text-align: center;">9</td>
              <td rowspan="1" colspan="1" col-key="j" class="fan-table-body-td fan-table-fixed-right" style="text-align: center; right: 0px;">10</td>
            </tr>
          </tbody>
          <tfoot class="fan-table-fixed-footer fan-table-footer"></tfoot>
        </table>
      </div>
    </div>
  </div>
</div>
`;
